File: src/tools/glob.ts
import { glob as globFn } from "glob";
import type { ToolResult } from "./bash.js";

const MAX_RESULTS = 10_000;

export interface GlobParams {
  pattern: string;
}

export async function glob(params: GlobParams): Promise<ToolResult> {
  try {
    const files = await globFn(params.pattern, {
      absolute: true,
      nodir: false,
      dot: true,
      ignore: ["**/node_modules/**", "**/.git/**"],
    });

    if (files.length === 0) {
      return { content: "No files matched the pattern." };
    }

    const truncated = files.length > MAX_RESULTS;
    const result = truncated ? files.slice(0, MAX_RESULTS) : files;
    let output = result.join("\n");
    if (truncated) {
      output += `\n\n(results truncated at ${MAX_RESULTS} of ${files.length} matches — use a more specific pattern)`;
    }

    return { content: output };
  } catch (err: any) {
    return {
      content: `Error: ${err.message}`,
      isError: true,
    };
  }
}

File: pincer/tools/glob.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const maxGlobResults = 10000

type GlobParams struct {
	Pattern string `json:"pattern"`
}

func Glob(p *GlobParams) (*Result, error) {
	// filepath.Glob doesn't support ** so we implement it with Walk
	if strings.Contains(p.Pattern, "**") {
		return globDoublestar(p.Pattern)
	}

	matches, err := filepath.Glob(p.Pattern)
	if err != nil {
		return &Result{Content: "Error: " + err.Error(), IsError: true}, nil
	}

	if len(matches) == 0 {
		return &Result{Content: "No files matched the pattern."}, nil
	}

	truncated := len(matches) > maxGlobResults
	if truncated {
		matches = matches[:maxGlobResults]
	}
	output := strings.Join(matches, "\n")
	if truncated {
		output += fmt.Sprintf("\n\n(results truncated at %d matches — use a more specific pattern)", maxGlobResults)
	}

	return &Result{Content: output}, nil
}

func globDoublestar(pattern string) (*Result, error) {
	// Split on ** to get prefix and suffix
	parts := strings.SplitN(pattern, "**", 2)
	root := filepath.Clean(parts[0])
	suffix := ""
	if len(parts) > 1 {
		suffix = strings.TrimPrefix(parts[1], "/")
		suffix = strings.TrimPrefix(suffix, string(filepath.Separator))
	}

	if root == "" || root == "." {
		root, _ = os.Getwd()
	}

	var matches []string
	truncated := false

	filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if len(matches) >= maxGlobResults {
			truncated = true
			return filepath.SkipAll
		}

		base := filepath.Base(path)
		if info.IsDir() && (base == ".git" || base == "node_modules") {
			return filepath.SkipDir
		}

		if suffix == "" {
			matches = append(matches, path)
			return nil
		}

		matched, _ := filepath.Match(suffix, filepath.Base(path))
		if matched {
			matches = append(matches, path)
		}

		return nil
	})

	if len(matches) == 0 {
		return &Result{Content: "No files matched the pattern."}, nil
	}

	output := strings.Join(matches, "\n")
	if truncated {
		output += fmt.Sprintf("\n\n(results truncated at %d matches — use a more specific pattern)", maxGlobResults)
	}

	return &Result{Content: output}, nil
}
